spec.proto

```diff
@@ -0,0 +1,55 @@
+/**
+ * CloudEvent Protobuf Format
+ *
+ * - Required context attributes are explicity represented.
+ * - Optional and Extension context attributes are carried in a map structure.
+ * - Data may be represented as binary, text, or protobuf messages.
+ */
+
+syntax = "proto3";
+
+package io.cloudevents.v1;
+
+import "google/protobuf/any.proto";
+import "google/protobuf/timestamp.proto";
+
+option java_package = "io.cloudevents.v1.proto";
+
+message CloudEvent {
+
+  // -- CloudEvent Context Attributes
+
+  // Required
+  string id = 1;
+  string source = 2; // URI-reference
+  string spec_version = 3;
+  string type = 4;
+
+  // Optional & Extension
+  map<string, CloudEventAttribute> attributes = 5;
+
+  // -- CloudEvent Data (Bytes, Text, or Proto)
+  oneof data_oneof {
+    bytes binary_data = 6;
+    string text_data = 7;
+    google.protobuf.Any proto_data = 8;
+  }
+
+  /**
+   * The CloudEvent specification defines
+   * seven atrribute types...
+   */
+
+  message CloudEventAttribute {
+
+    oneof attr_oneof {
+      bool cloud_event_boolean = 1;
+      sfixed32 cloud_event_integer = 2;
```

Comment:
Not sure if you need/want `sfixed32` here -- this is a fixed-length encoding rather than a varint, which is typical. This means it will always be 4 bytes long, rather than being 1-5 bytes long depending on content. If integers follow a power law distribution, `int32` will almost certainly be more compact; if you have negative integers more than about 20% of the time, then `sint32` will probably be more compact.

Reply:
Ack .. I think i was going for _predictability_ at the time, _efficiency_ is probably a better driver..